docs/DevGuide/TensorExpressions.md

```diff
@@ -0,0 +1,411 @@
+\cond NEVER
+Distributed under the MIT License.
+See LICENSE.txt for details.
+\endcond
+# Writing tensor equations with TensorExpressions {#writing_tensorexpressions}
+
+\tableofcontents
+
+SpECTRE's `TensorExpression`s interface allows you to write tensor equations in
+SpECTRE in C++ with syntax that resembles tensor index notation. To use it,
+simply add this include to the top of your file:
+```
+#include "DataStructures/Tensor/Tensor.hpp"
+```
+The following guide assumes a basic understanding of the `Tensor` class and
+\ref tnsr "tnsr" type aliases.
+
+# Syntax {#te_syntax}
+`TensorExpression`s are arithmetic expressions of `Tensor`s that can be
+evaluated using `tenex::evaluate`. Terms used in the expression may be `Tensor`s
+or numbers (see [supported types](#te_data_type_support)).
+
+As a simple example of how `TensorExpression`s are used, if you would like to
+raise the index of some `Tensor` `R` with some spacetime metric `Tensor` `g`,
+i.e. \f$R^c{}_b = R_{ab} g^{ac}\f$, you can compute this with
+`TensorExpression`s by doing:
+```
+auto R_up = tenex::evaluate<ti::C, ti::b>(R(ti::a, ti::b) * g(ti::A, ti::C));
+```
+
+where `R_up`, `R`, and `g` are rank 2 spacetime `Tensor`s and the `ti::*`
+variables are `TensorIndex`s. Here, the argument to
+\ref tenex::evaluate "evaluate" is the RHS tensor expression to compute, `R_up`
+is the result LHS `Tensor`, and the template arguments to
+\ref tenex::evaluate "evaluate" are the LHS `Tensor`'s indices. The LHS symmetry
+will be deduced from the RHS tensors' symmetries and order of operations.
+
+Alternatively, if you already have a LHS `Tensor` variable, you can pass it into
+the following \ref tenex::evaluate "evaluate" overload, where the LHS `Tensor`
+provided will be assigned to the result of the RHS expression:
+
+```
+tenex::evaluate<ti::C, ti::b>(
+    make_not_null(&R_up), R(ti::a, ti::b) * g(ti::A, ti::C));
+```
+
+One advantage of this overload is that it uses the symmetry and index types
+(spatial/spacetime) of the provided LHS tensor instead of deducing them from the
+RHS expression, so this enables you to specify the LHS index structure in cases
+where the previous \ref tenex::evaluate "evaluate" overload does not deduce the
+one you want. Note that the LHS tensor does not need to be previously sized
```

Comment:
If would occur in something as simple as this:
```
tnsr::i<double, 3> R{};
auto L = tenex::evaluate<ti::a, ti::b>(R(ti::a) * R(ti::b));
```
because the compiler (result symmetry is computed at compile time) doesn't know that the two operands to `*` are the same `Tensor`. `L` would be deduced as `tnsr::ij`, not `tnsr::ii`.

I think it's useful to demonstrate this, and the only reason I didn't is because when I was writing this, I was trying to present meat of how to write an equation at the top of the file without writing lots of details and getting into the weeds until down below. Like "here is what it looks like, now let me tell you more below." But you're someone who experienced reading it so, do you think it would be best to include an example like the one I showed up at the top in the `Syntax` section where I'm introducing the two `evaluate` overloads, or do you think it's best to allude to it and link it to an example further down like I did with other things?

Reply:
I'd say link to an example further down. I like your approach of using simple examples at the top to just demonstrate things and then having more detailed examples below.